<a href="https://colab.research.google.com/github/rachna7063/My-_Project-_work/blob/main/Part_1_Data_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Stream_Flix Project**


---







In [2]:
print("Welcome, to my Project ")

Welcome, to my Project 


Data cleaning process is starting from here,,


In [3]:
import os   # I am importing libraries , in google colab no need to install (!pip install pandas)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
ratings_data = pd.read_csv("/content/ratings.csv")

In [5]:
ratings_data

,rating_id,subscriber_id,title_id,rating,rating_date
0,1,SUB109184,TTL203909,4,2026-02-06
1,2,SUB105194,TTL200888,4,2025-03-21
2,3,SUB102974,TTL200036,3,2024-05-26
3,4,SUB105612,TTL206358,5,2026-05-22
4,5,SUB103070,TTL205227,4,2026-05-16
...,...,...,...,...,...
129995,129996,SUB110434,TTL207853,4,2025-11-26
129996,129997,SUB100284,TTL207155,5,2023-08-08
129997,129998,SUB109073,TTL204687,5,2024-12-02
129998,129999,SUB104448,TTL205597,3,2024-07-02


In [6]:
ratings_data.shape

(130000, 5)

In [7]:
ratings_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130000 entries, 0 to 129999
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   rating_id      130000 non-null  int64 
 1   subscriber_id  130000 non-null  object
 2   title_id       130000 non-null  object
 3   rating         130000 non-null  int64 
 4   rating_date    130000 non-null  object
dtypes: int64(2), object(3)
memory usage: 5.0+ MB


In [8]:
ratings_data["rating_date"] = pd.to_datetime(ratings_data["rating_date"])

here datetime is showing as object so i have to change it in proper date_time format.  


---

so after coding i will check again that is it in proper format or not.

In [9]:
ratings_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130000 entries, 0 to 129999
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   rating_id      130000 non-null  int64         
 1   subscriber_id  130000 non-null  object        
 2   title_id       130000 non-null  object        
 3   rating         130000 non-null  int64         
 4   rating_date    130000 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(2), object(2)
memory usage: 5.0+ MB


In [10]:
ratings_data.isnull().sum()

,0
rating_id,0
subscriber_id,0
title_id,0
rating,0
rating_date,0


After reviewing it carefull there is not any null value in dataset.


In [11]:
ratings_data.duplicated().sum()

np.int64(0)

I have checked for duplicate value but here is not any duplicate value.

In [12]:
 #1. remove Ratings(if rating 5 is greater than 5 or smaller than 1 )
ratings_data = ratings_data[(ratings_data['rating'] >= 1) & (ratings_data['rating'] <= 5)]

In [13]:
print("Minimum Rating is in dataset:", ratings_data['rating'].min())
print("Maximum Rating is in dataset:", ratings_data['rating'].max())

Minimum Rating is in dataset: 1
Maximum Rating is in dataset: 5


so rating is in range 1 to 5

NOW i am moving towards loading another dataset i.e. **reviews_data**

In [14]:
print("Read csv files")
reviews_data= pd.read_csv("/content/reviews.csv")

Read csv files


In [15]:
reviews_data

,review_id,subscriber_id,title_id,review_text,sentiment,helpful_votes,review_date
0,1,SUB102409,TTL206425,Stellar cast and a satisfying ending.,Positive,9,2023-11-04
1,2,SUB111270,TTL203987,Predictable plot and flat characters.,Negative,4,2026-04-25
2,3,SUB107848,TTL203235,"Started strong, lost steam in the middle.",Neutral,2,2024-07-31
3,4,SUB114501,TTL200930,Beautifully shot and emotionally resonant.,Positive,14,2022-12-13
4,5,SUB113370,TTL207141,A rare gem. Highly recommend.,Positive,6,2025-10-09
...,...,...,...,...,...,...,...
109995,109996,SUB100390,TTL207141,Average. Worth it if you like the genre.,Neutral,3,2025-05-31
109996,109997,SUB114968,TTL208918,Disappointing given the hype.,Negative,4,2025-12-03
109997,109998,SUB112680,TTL205235,A rare gem. Highly recommend.,Positive,6,2024-10-05
109998,109999,SUB107694,TTL202380,"Good for a quiet evening, fairly forgettable.",Neutral,6,2026-05-18


In [16]:
reviews_data.shape

(110000, 7)

reviews_data has 110000 rows and 7 columns

In [17]:
reviews_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110000 entries, 0 to 109999
Data columns (total 7 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   review_id      110000 non-null  int64 
 1   subscriber_id  110000 non-null  object
 2   title_id       110000 non-null  object
 3   review_text    110000 non-null  object
 4   sentiment      110000 non-null  object
 5   helpful_votes  110000 non-null  int64 
 6   review_date    110000 non-null  object
dtypes: int64(2), object(5)
memory usage: 5.9+ MB


here datetime is in object data type so i have to change it in datetime data type

In [18]:
reviews_data["review_date"] = pd.to_datetime(reviews_data["review_date"])

In [19]:
reviews_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110000 entries, 0 to 109999
Data columns (total 7 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   review_id      110000 non-null  int64         
 1   subscriber_id  110000 non-null  object        
 2   title_id       110000 non-null  object        
 3   review_text    110000 non-null  object        
 4   sentiment      110000 non-null  object        
 5   helpful_votes  110000 non-null  int64         
 6   review_date    110000 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(2), object(4)
memory usage: 5.9+ MB


so i have changed review_date as object to datetime format

In [20]:
reviews_data.isnull().sum()

,0
review_id,0
subscriber_id,0
title_id,0
review_text,0
sentiment,0
helpful_votes,0
review_date,0


Here is not any null values

In [21]:
print("--- Unique Sentiments ---")
print(reviews_data['sentiment'].value_counts())

--- Unique Sentiments ---
sentiment
Positive    75007
Neutral     22759
Negative    12234
Name: count, dtype: int64


I have checked that how many unique sentiments are there so i found that
positive sentiments -75007,
negative sentiment-12234,
neutral sentiment -22759, so here positive sentiments is almost (approx 68%),negative sentiments(approx 11%)


In [22]:
print("\n--- Helpful Votes Statistics ---")
print("Minimum Votes:", reviews_data['helpful_votes'].min())
print("Maximum Votes:", reviews_data['helpful_votes'].max())


--- Helpful Votes Statistics ---
Minimum Votes: 0
Maximum Votes: 24


Here no any negative votings are there

Next move to load subscriber _data

In [23]:
subscribers_data = pd.read_csv("/content/subscribers.csv")

In [24]:
subscribers_data

,subscriber_id,signup_date,country,region,age,gender,plan_type,monthly_price_usd,household_size,primary_device,payment_method,tenure_months,is_active,churn_date
0,SUB100000,2023-01-09,France,Europe,22,Female,Basic with Ads,6.99,1,Mobile,Credit Card,41,True,NaN
1,SUB100001,2020-02-05,United Kingdom,Europe,48,Male,Basic with Ads,6.99,1,Smart TV,Credit Card,76,False,2022-12-08
2,SUB100002,2022-04-15,Italy,Europe,35,Male,Standard,15.49,2,Streaming Stick,Gift Card,50,True,NaN
3,SUB100003,2024-10-10,Spain,Europe,49,Female,Basic with Ads,6.99,1,Laptop,Credit Card,19,True,NaN
4,SUB100004,2020-04-12,United States,North America,36,Female,Basic with Ads,6.99,1,Mobile,Credit Card,74,False,2021-12-15
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14995,SUB114995,2023-06-14,Italy,Europe,40,Male,Standard,15.49,2,Mobile,PayPal,36,True,NaN
14996,SUB114996,2025-11-29,Japan,East Asia,54,Female,Standard,15.49,1,Tablet,PayPal,6,False,2026-03-23
14997,SUB114997,2025-09-16,United States,North America,22,Male,Basic with Ads,6.99,1,Smart TV,Credit Card,8,True,NaN
14998,SUB114998,2023-06-04,India,South Asia,19,Male,Basic with Ads,6.99,1,Tablet,PayPal,36,True,NaN


In [25]:
subscribers_data.shape

(15000, 14)

subscribers_data has 1500 rows and 14 columns

In [26]:
subscribers_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   subscriber_id      15000 non-null  object 
 1   signup_date        15000 non-null  object 
 2   country            15000 non-null  object 
 3   region             15000 non-null  object 
 4   age                15000 non-null  int64  
 5   gender             15000 non-null  object 
 6   plan_type          15000 non-null  object 
 7   monthly_price_usd  15000 non-null  float64
 8   household_size     15000 non-null  int64  
 9   primary_device     15000 non-null  object 
 10  payment_method     15000 non-null  object 
 11  tenure_months      15000 non-null  int64  
 12  is_active          15000 non-null  bool   
 13  churn_date         3801 non-null   object 
dtypes: bool(1), float64(1), int64(3), object(9)
memory usage: 1.5+ MB


signuo_date and churn_date both are in object data type so ,  i have to change it right data type.

In [27]:
# To change date columns in rigt format  (Datetime)
subscribers_data['signup_date'] = pd.to_datetime(subscribers_data['signup_date'], errors='coerce')
subscribers_data['churn_date'] = pd.to_datetime(subscribers_data['churn_date'], errors='coerce')


In [28]:
subscribers_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   subscriber_id      15000 non-null  object        
 1   signup_date        15000 non-null  datetime64[ns]
 2   country            15000 non-null  object        
 3   region             15000 non-null  object        
 4   age                15000 non-null  int64         
 5   gender             15000 non-null  object        
 6   plan_type          15000 non-null  object        
 7   monthly_price_usd  15000 non-null  float64       
 8   household_size     15000 non-null  int64         
 9   primary_device     15000 non-null  object        
 10  payment_method     15000 non-null  object        
 11  tenure_months      15000 non-null  int64         
 12  is_active          15000 non-null  bool          
 13  churn_date         3801 non-null   datetime64[ns]
dtypes: boo

In [29]:
subscribers_data.duplicated().sum()

np.int64(0)

so here is no any duplicate value in subscribers_data

In [30]:
subscribers_data.isnull().sum()

,0
subscriber_id,0
signup_date,0
country,0
region,0
age,0
gender,0
plan_type,0
monthly_price_usd,0
household_size,0
primary_device,0


IN churn_date - 11,199 null values .That means at platform  11,199 users are still Active ! (Total 15,000 users - 11,199 active = 3,801 churned users).

CHECK 1: Confirm churn_date is after signup_date
# =========================================================================


# A New column 'flag_invalid_churn_date' (True = wrong data)

In [31]:

subscribers_data['flag_invalid_churn_date'] = (
    subscribers_data['churn_date'].notna() &
    (subscribers_data['churn_date'] < subscribers_data['signup_date'])
)



In [32]:
invalid_dates_count = subscribers_data['flag_invalid_churn_date'].sum()
print(f"confirm: Total {invalid_dates_count} rows are there where churn_dateis before signup_date .")


confirm: Total 0 rows are there where churn_dateis before signup_date .


CHECK 2: Confirm active subscribers have a blank churn_date

I will create a column 'flag_active_with_churn_date' (True = Mismatch)

In [33]:
subscribers_data['flag_active_with_churn_date'] = (
    (subscribers_data['is_active'] == True) &
    subscribers_data['churn_date'].notna()
)


In [34]:
mismatch_active_count = subscribers_data['flag_active_with_churn_date'].sum()
print(f" Confirm active : Total {mismatch_active_count} rows are found where subscriber are active but written churn date .")

 Confirm active : Total 0 rows are found where subscriber are active but written churn date .


now move to upload another data i.e . title_data

In [35]:
titles_data = pd.read_csv("/content/titles.csv")

In [36]:
titles_data.head()

,title_id,title_name,type,primary_genre,country,language,release_year,date_added,maturity_rating,seasons,...,is_original,license_type,director,cast,quality_score,popularity_score,license_cost_usd,license_expiry,total_watch_hours,total_plays
0,TTL200000,Toronto Requiem,TV Show,Comedy,South Korea,Korean,2026,2026-04-05,R,6,...,True,Original,Ravi Nguyen,"Rosa Ferrari, Aria Cohen, Nina Kim, Felix Santos",68.6,0.19,7965521,NaN,1300.8,44
1,TTL200001,Seoul Inheritance,TV Show,Fantasy,South Korea,Korean,2026,2026-05-01,TV-MA,4,...,False,Non-Exclusive License,Greta Sharma,"Priya Ferrari, Lucia Mehta, Ravi Yilmaz, Joon ...",89.4,0.29,3074516,2028-08-03,1767.2,52
2,TTL200002,Tender Paradox,TV Show,Horror,Italy,Italian,2026,2026-04-17,PG,3,...,False,Non-Exclusive License,Bruno Moreau,"Elena Park, Aria Romano, Omar Olsen, Chloe Nak...",37.3,0.20,629307,2027-02-16,166.5,60
3,TTL200003,Velvet Expedition,Movie,Animation,India,Hindi,1985,2018-12-08,TV-14,0,...,False,Non-Exclusive License,Owen Mwangi,"Ava Russo, Clara Fischer, Zoe Khan, Rosa Reyes",37.8,0.02,1345769,2029-10-11,16.2,43
4,TTL200004,The Shadows of Lisbon,Movie,Animation,India,Hindi,1991,2016-05-11,TV-14,0,...,False,Exclusive License,Joon Kim,"Hana Patel, Fatima Patel, Hassan Reyes, Liam L...",78.0,0.00,1696493,2028-11-21,35.7,43


In [37]:
titles_data.shape

(9000, 21)

titles_data contains 9000 rows and 21 columns .

In [38]:
titles_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   title_id              9000 non-null   object 
 1   title_name            9000 non-null   object 
 2   type                  9000 non-null   object 
 3   primary_genre         9000 non-null   object 
 4   country               9000 non-null   object 
 5   language              9000 non-null   object 
 6   release_year          9000 non-null   int64  
 7   date_added            9000 non-null   object 
 8   maturity_rating       9000 non-null   object 
 9   seasons               9000 non-null   int64  
 10  content_duration_min  9000 non-null   int64  
 11  is_original           9000 non-null   bool   
 12  license_type          9000 non-null   object 
 13  director              9000 non-null   object 
 14  cast                  9000 non-null   object 
 15  quality_score        

In [39]:
#  to change datetime in right format
titles_data['date_added'] = pd.to_datetime(titles_data['date_added'], errors='coerce')
titles_data['license_expiry'] = pd.to_datetime(titles_data['license_expiry'], errors='coerce')


In [40]:
titles_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   title_id              9000 non-null   object        
 1   title_name            9000 non-null   object        
 2   type                  9000 non-null   object        
 3   primary_genre         9000 non-null   object        
 4   country               9000 non-null   object        
 5   language              9000 non-null   object        
 6   release_year          9000 non-null   int64         
 7   date_added            9000 non-null   datetime64[ns]
 8   maturity_rating       9000 non-null   object        
 9   seasons               9000 non-null   int64         
 10  content_duration_min  9000 non-null   int64         
 11  is_original           9000 non-null   bool          
 12  license_type          9000 non-null   object        
 13  director          

In [41]:
 # standardise the columns ( remove Extra spaces )
# Select only the columns that actually contain string/object data
text_cols = titles_data.select_dtypes(include=['object']).columns

for col in text_cols:
    titles_data[col] = titles_data[col].astype(str).str.strip().str.title()

In [42]:
print(titles_data['primary_genre'].unique())


['Comedy' 'Fantasy' 'Horror' 'Animation' 'Action' 'Crime' 'Drama'
 'Documentary' 'Mystery' 'Thriller' 'Sci-Fi' 'Romance' 'Kids & Family'
 'Reality' 'Stand-Up' 'Musical']


In [43]:
print(titles_data["type"].unique())

['Tv Show' 'Movie']


Now I am uploading **watch_history_data**

In [44]:
watch_history_data= pd.read_csv("/content/watch_history.csv")

In [45]:
watch_history_data

,watch_id,subscriber_id,title_id,watch_date,device,region,content_duration_min,watch_duration_min,completion_pct,completed
0,1,SUB107630,TTL205207,2024-01-01,Smart TV,East Asia,888,826.7,93.1,True
1,2,SUB114773,TTL204889,2025-10-18,Mobile,South Asia,2340,1866.4,79.8,False
2,3,SUB106013,TTL206882,2026-04-26,Mobile,Latin America,656,75.0,11.4,False
3,4,SUB100939,TTL208849,2026-03-09,Mobile,North America,500,493.2,98.6,True
4,5,SUB107223,TTL202529,2023-12-27,Tablet,North America,141,94.6,67.1,False
...,...,...,...,...,...,...,...,...,...,...
649995,649996,SUB113946,TTL207486,2025-07-13,Mobile,Europe,444,8.6,1.9,False
649996,649997,SUB108172,TTL200072,2026-05-22,Smart TV,Latin America,420,338.3,80.5,False
649997,649998,SUB108092,TTL204011,2025-01-25,Streaming Stick,South Asia,261,47.4,18.2,False
649998,649999,SUB114820,TTL204762,2026-04-13,Tablet,North America,77,11.6,15.0,False


In [46]:
watch_history_data.shape

(650000, 10)

In [47]:
watch_history_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 650000 entries, 0 to 649999
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   watch_id              650000 non-null  int64  
 1   subscriber_id         650000 non-null  object 
 2   title_id              650000 non-null  object 
 3   watch_date            650000 non-null  object 
 4   device                650000 non-null  object 
 5   region                650000 non-null  object 
 6   content_duration_min  650000 non-null  int64  
 7   watch_duration_min    650000 non-null  float64
 8   completion_pct        650000 non-null  float64
 9   completed             650000 non-null  bool   
dtypes: bool(1), float64(2), int64(2), object(5)
memory usage: 45.3+ MB


In [48]:
print("--- Watch History Data Cleaning Shuru ---")

#  Convert Date column into Datetime format
watch_history_data['watch_date'] = pd.to_datetime(watch_history_data['watch_date'], errors='coerce')


--- Watch History Data Cleaning Shuru ---


In [49]:
print("---check it that watch_date column converted in date time format")
watch_history_data.info()

---check it that watch_date column converted in date time format
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 650000 entries, 0 to 649999
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   watch_id              650000 non-null  int64         
 1   subscriber_id         650000 non-null  object        
 2   title_id              650000 non-null  object        
 3   watch_date            650000 non-null  datetime64[ns]
 4   device                650000 non-null  object        
 5   region                650000 non-null  object        
 6   content_duration_min  650000 non-null  int64         
 7   watch_duration_min    650000 non-null  float64       
 8   completion_pct        650000 non-null  float64       
 9   completed             650000 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(2), int64(2), object(4)
memory usage: 45.3+ MB


**successfull**

In [50]:
#  To clean aur standardize  the text column
text_cols = ['device', 'region']
for col in text_cols:
    watch_history_data[col] = watch_history_data[col].astype(str).str.strip().str.title()

In [51]:
#  Outlier Fix: remove negative value
watch_history_data = watch_history_data[watch_history_data['watch_duration_min'] >= 0]
watch_history_data = watch_history_data[watch_history_data['content_duration_min'] > 0]


In [52]:
#CHECK 1: Confirm Watch Duration is less than or equal to Content Duration
# =========================================================================
# if watch_duration > content_duration than entry (mismatch)
watch_history_data['flag_invalid_duration'] = watch_history_data['watch_duration_min'] > watch_history_data['content_duration_min']
mismatch_duration = watch_history_data['flag_invalid_duration'].sum()
print(f" Check 1: Total {mismatch_duration} rows finding where watch duration is greater than content duration.")

 Check 1: Total 0 rows finding where watch duration is greater than content duration.


In [53]:
#CHECK 2: TO  Recalculate the completion percentage
# =========================================================================
# TO apply formula for the right calculation
watch_history_data['completion_percentage'] = (watch_history_data['watch_duration_min'] / watch_history_data['content_duration_min']) * 100

print("\n Watch History table is completely clean aur logically validated!")




 Watch History table is completely clean aur logically validated!


In [54]:
watch_history_data['completion_percentage']

,completion_percentage
0,93.096847
1,79.760684
2,11.432927
3,98.640000
4,67.092199
...,...
649995,1.936937
649996,80.547619
649997,18.160920
649998,15.064935


In [55]:
watch_history_data['completed'] = np.where(watch_history_data['completion_pct'] >= 90, True, False)

print("--- Completion Checks Successfully Updated! ---")
print("Completion Percentage (Min):", watch_history_data['completion_pct'].min(), "%")
print("Completion Percentage (Max):", watch_history_data['completion_pct'].max(), "%")
print("\nVideo Completion Counts:")
print(watch_history_data['completed'].value_counts())

--- Completion Checks Successfully Updated! ---
Completion Percentage (Min): 1.0 %
Completion Percentage (Max): 100.0 %

Video Completion Counts:
completed
False    550102
True      99898
Name: count, dtype: int64


Now i am uploading **watchlist_data**

In [56]:
watchlist_data = pd.read_csv("/content/watchlist.csv")

In [57]:
watchlist_data

,watchlist_id,subscriber_id,title_id,added_date,watched
0,1,SUB104004,TTL201827,2023-06-05,True
1,2,SUB102310,TTL201354,2023-01-31,False
2,3,SUB111860,TTL204871,2022-06-21,True
3,4,SUB113333,TTL205130,2023-12-23,True
4,5,SUB106850,TTL204396,2022-02-02,False
...,...,...,...,...,...
64995,64996,SUB110813,TTL201351,2025-03-28,True
64996,64997,SUB108906,TTL207032,2026-02-02,False
64997,64998,SUB113557,TTL200478,2026-05-22,False
64998,64999,SUB108779,TTL207655,2024-12-21,False


In [58]:
watchlist_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65000 entries, 0 to 64999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   watchlist_id   65000 non-null  int64 
 1   subscriber_id  65000 non-null  object
 2   title_id       65000 non-null  object
 3   added_date     65000 non-null  object
 4   watched        65000 non-null  bool  
dtypes: bool(1), int64(1), object(3)
memory usage: 2.0+ MB


In [59]:
print("\n-----convert added_date column data type into datetime data type")
watchlist_data['added_date'] = pd.to_datetime(watchlist_data['added_date'], errors='coerce')


-----convert added_date column data type into datetime data type


In [60]:
#check added_data Dtype
watchlist_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65000 entries, 0 to 64999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   watchlist_id   65000 non-null  int64         
 1   subscriber_id  65000 non-null  object        
 2   title_id       65000 non-null  object        
 3   added_date     65000 non-null  datetime64[ns]
 4   watched        65000 non-null  bool          
dtypes: bool(1), datetime64[ns](1), int64(1), object(2)
memory usage: 2.0+ MB


In [61]:
 #Clean the Boolean column 'watched'
watchlist_data['watched'] = watchlist_data['watched'].astype(bool)

In [62]:
watchlist_data.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
64995,False
64996,False
64997,False
64998,False


here finding is no any duplicate values has found but again i will check by using subset  by combining  subscriber and Title

In [63]:
#To Check that in Subscriber and  Title combination, any duplicates value is there
sub_title_duplicates = watchlist_data.duplicated(subset=['subscriber_id', 'title_id']).sum()

print(f"duplicate rows in Subscriber and Title combination: {sub_title_duplicates}")

duplicate rows in Subscriber and Title combination: 61


ohh!  Here is 61 duplicate rows in subscriber and titile combination. so, i will drop all the duplicate value to gain exact insight

In [64]:
#Remove all the duplicate rows
watchlist_data.drop_duplicates(subset=['subscriber_id', 'title_id'], keep='first', inplace=True)
#indicator flag
watchlist_data['in_watchlist'] = 1

print("✔ 61 Duplicate rows deleted successfully !")
print(f"total remaining rows in watchlist table: {len(watchlist_data):,}")

✔ 61 Duplicate rows deleted successfully !
total remaining rows in watchlist table: 64,939


Earlier there were 65000 rows but after deleting the dulicates rows it remains only 64,939 Rows

In [65]:
print("\n----ratings_data")
ratings_data.shape



----ratings_data


(130000, 5)

In [66]:
print("\n---reviews_data")
reviews_data.shape



---reviews_data


(110000, 7)

In [67]:
print("\n---subscribers_data")
subscribers_data.shape



---subscribers_data


(15000, 16)

In [68]:
print("\n---titles_data")
titles_data.shape



---titles_data


(9000, 21)

In [69]:
print("\n---watch_history_data")
watch_history_data.shape



---watch_history_data


(650000, 12)

In [70]:
print("\n---watchlist_data")
watchlist_data.shape


---watchlist_data


(64939, 6)

In [71]:
subscribers_data.columns = subscribers_data.columns.str.strip().str.lower()
titles_data.columns = titles_data.columns.str.strip().str.lower()


In [72]:
# To avoid the KeyError, only cast columns that actually exist in the DataFrame:
# subscribers_data does not have a 'title_id' column, but titles_data does.
titles_data['title_id'] = titles_data['title_id'].astype(str)

In [73]:
# To link subscribers to titles, we use 'watch_history_data' as a bridge table
temp_df = pd.merge(watch_history_data, subscribers_data, on='subscriber_id', how='left')
final_data = pd.merge(temp_df, titles_data, on='title_id', how='left')

In [74]:
final_data

,watch_id,subscriber_id,title_id,watch_date,device,region_x,content_duration_min_x,watch_duration_min,completion_pct,completed,...,is_original,license_type,director,cast,quality_score,popularity_score,license_cost_usd,license_expiry,total_watch_hours,total_plays
0,1,SUB107630,TTL205207,2024-01-01,Smart Tv,East Asia,888,826.7,93.1,True,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN
1,2,SUB114773,TTL204889,2025-10-18,Mobile,South Asia,2340,1866.4,79.8,False,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN
2,3,SUB106013,TTL206882,2026-04-26,Mobile,Latin America,656,75.0,11.4,False,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN
3,4,SUB100939,TTL208849,2026-03-09,Mobile,North America,500,493.2,98.6,True,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN
4,5,SUB107223,TTL202529,2023-12-27,Tablet,North America,141,94.6,67.1,False,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
649995,649996,SUB113946,TTL207486,2025-07-13,Mobile,Europe,444,8.6,1.9,False,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN
649996,649997,SUB108172,TTL200072,2026-05-22,Smart Tv,Latin America,420,338.3,80.5,False,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN
649997,649998,SUB108092,TTL204011,2025-01-25,Streaming Stick,South Asia,261,47.4,18.2,False,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN
649998,649999,SUB114820,TTL204762,2026-04-13,Tablet,North America,77,11.6,15.0,False,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN


In [75]:
watchlist_data.columns = watchlist_data.columns.str.strip().str.lower()
reviews_data.columns = reviews_data.columns.str.strip().str.lower()

In [76]:
watchlist_data['title_id'] = watchlist_data['title_id'].astype(str)